pytorch_lightning/trainer/connectors/checkpoint_connector.py

```diff
@@ -315,7 +316,10 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
                 'epoch':                     training epoch
                 'global_step':               training global step
                 'pytorch-lightning_version': The version of PyTorch Lightning that produced this checkpoint
-                'callbacks':                 "callback specific state"[] # if not weights_only
+                'callbacks_state_dict':                                  # if not weights_only
+                    {callback.state_key: callback.state_dict}
+                'callbacks_deprecated_hook_states':                      # if not weights_only
+                    {callback.state_key: callback.deprecated_on_save_checkpoint}
```

Comment:
we cannot break backwards compatibility with checkpoints saved by prior versions of the framework. changing the dict structure here does that

Reply:
I would much prefer a path that allows incremental adoption of state_dict and load_state_dict instead while preserving checkpoint backward compatibility.

To that end, if users return something in Callback.on_save_checkpoint, why not let it override Callback.state_dict during the deprecation phase? It's highly unlikely users will have implementations for both, given state_dict is brand new. This lets people cut over without needing to change the serialized checkpoint structure. We can make this temporary affordance between the checkpoint connector/trainer logic.